types/eslint-utils/index.d.ts

```diff
@@ -0,0 +1,217 @@
+
+export type StaticValue = StaticValueProvided | StaticValueOptional;
+
+export interface StaticValueProvided {
+    value: any;
+}
+
+export interface StaticValueOptional {
+    optional?: true;
+    value: undefined;
+}
+
+/**
+ * Get the value of a given node if it's a static value.
+ * @param node The node to get.
+ * @param [initialScope] The scope to start finding variable. Optional. If this scope was given, this tries to resolve identifier references which are in the given node as much as possible.
+ * @returns The static value of the node, or `null`.
+ */
+export function getStaticValue(node: Node, initialScope?: Scope.Scope | null): StaticValue | null;
+
+/**
+ * Get the value of a given node if it's a literal or a template literal.
+ * @param node The node to get.
+ * @param initialScope The scope to start finding variable.
+ * @returns The value of the node, or `null`.
+ * @remarks
+ * If the node is an Identifier node and scope was given, this checks the variable of the identifier,
+ * and returns the value of it if the variable is a constant.
+ */
+export function getStringIfConstant(node: Node, initialScope?: Scope.Scope | null): string | null;
+
+/**
+ * Options for `hasSideEffect`, optionally.
+ */
+export interface HasSideEffectOptions {
+    /**
+     * If `true` then it considers member accesses as the node which has side effects.
+     */
+    considerGetters?: boolean;
+
+    /**
+     * If `true` then it considers implicit type conversion as the node which has side effects.
+     */
+    considerImplicitTypeConversion?: boolean;
+
+    /**
+     * The keys to traverse nodes. Use `context.getSourceCode().visitorKeys`
+     */
+    visitorKeys?: VisitorKeys;
+}
+
+/**
+ * Check whether a given node has any side effect or not.
+ * @param node The node to get.
+ * @param sourceCode The source code object.
+ * @param options The option object.
+ * @returns `true` if the node has a certain side effect.
+ */
+export function hasSideEffect(node: Node, sourceCode: SourceCode, options?: HasSideEffectOptions): boolean;
+
+/**
+ * Check whether a given node is parenthesized or not.
+ * @param times The number of parantheses.
+ * @param node The AST node to check.
+ * @param sourceCode The source code object to get tokens.
+ * @returns `true` if the node is parenthesized the given times.
+ */
+export function isParenthesized(times: number, node: Node, sourceCode: SourceCode): boolean;
+/**
+ * Check whether a given node is parenthesized or not.
+ * @param node The AST node to check.
+ * @param sourceCode The source code object to get tokens.
+ * @returns `true` if the node is parenthesized.
+ */
+export function isParenthesized(node: Node, sourceCode: SourceCode): boolean;
+
+/**
+ * Checks if the given token is an arrow token or not.
+ * @param token The token to check.
+ * @returns `true` if the token is an arrow token.
+ */
+export function isArrowToken(token: AST.Token): boolean;
```

Comment:
I think any function that checks tokens can pass a `Comment` as a `token` argument.

```suggestion
export function isArrowToken(token: AST.Token | import('estree').Comment): boolean;
```

https://github.com/typescript-eslint/typescript-eslint/blob/1c1b572c3000d72cfe665b7afbada0ec415e7855/packages/experimental-utils/src/ast-utils/eslint-utils/predicates.ts#L4

Also, can you use `is` to make it a typeguard function?

```suggestion
export function isArrowToken(token: AST.Token | import('estree').Comment):  token is AST.Token & { type: 'Punctuator', value: '=>' };
```

Reply:
@ota-meshi I'm hesitant to add `& { ... }` types in an `is` typeguard. Do you think it'd be reasonable to separately add them to `@types/eslint` and then refer to them here

Reply:
I think it's better to add type to `@types/eslint`  before referencing it as you say.

Reply:
I've added a few interfaces for the `type: "Punctuator"` ones referenced in this file.

```ts
    interface PunctuatorToken<Value extends string> extends Token {
        type: 'Punctuator';
        value: Value;
    }

    interface ArrowToken extends PunctuatorToken<'=>'> {}

    ...
```

I'll note that the comma token is a bit different because its `"Block" | "Line" | "Shebang"` type isn't assignable to the existing `Token`'s `type`, so I've left that without its own specialized type and as a `boolean` here.